I1116 XML standings changes for FTS, medals and Bill Honors rules

johnbrvc opened the pull request:
### Description of what the PR does
Add `color `and `rgb `attributes to the `<problem> `tags in the XML standings generation.
Add first-to-solve (`fts`) attribute to `<problemSummaryInfo> `tags in the XML standings generation.
Add `isGold`, `isSilver`, `isBronze`, `isHighest`, `isHigh`, `isHonors`, `isHonorable` optional attributes to the `<teamStanding>` tags.  Each attribute will be set to true if the team qualifies for the citation.
Update the PC2 API to pass the fts attribute.
Add `elapsedtime `and `remainingtime `attributes to `<standingsHeader>`
Ignore new attributes in standings generation reports and **CLICS 2023-06 API**.
CI: Fixed interface description comments for `pc2/api/IProblemDetails.java`

Extensive testing with the WTI and report generation was tested to ensure that adding these new attributes does not break anything.  Logs were checked to make sure there were no exceptions generated.

The over goal of this PR is to allow the XSL template files to better format scoreboards and convey more information accurately. 

### Issue which the PR addresses
Fixes #1116 
### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

1. Start up a contest that has a reasonable scoreboard (many entries), such as NAC2025
2. Start up an administrator
3. On the **Finalize** tab, change medals to: 2G/2S/2B and check the "**Use World Finals group rankings for results**" box.
4. Apply the settings (**Update**)
5. Start a scoreboard client.
6. Look at the `results.xml `file generated in the pc2 directory.
7. It should have the new attributes in it as shown in the file attached.

[results_061526_fts_medals_honors.txt](https://github.com/user-attachments/files/20742020/results_061526_fts_medals_honors.txt)

johnbrvc committed:
i_1116 Add FTS, colors and medals to standings xml

Add fts attribute to each problemSummary tag.
Add isGold or isSilver or isBronze if a team should get that medal (standingsrecord)
Add color and rgb to each problem tag.
Ignore these attributes when converting the scoreboard to JSON.

johnbrvc committed:
i_1116 add flags for Bill rules for standings entries

Add FTS to IProblemDetails details as a property.
Update reports and JUnits to print out FTS
Change nomenclature for medal places as opposed to ranks.
In DSA, remember if we excluded groups in the standings (that is, someone only wants standings for one or a couple of groups).  In this case we won't generate the Bill rules or medals.
Need to remember the XML memento for each teamStanding wo we can fill in the medals and Bill rules after the whole thing is generated.
Clean up new method in ResultsFile to calculate the ranks for Bill Honor rules.
CI change string compare with constant to avoid possibility of NPE

johnbrvc committed:
i_1116 Fixed JUnit bug

Junits do not set contest times.  When elapsed/remaining time were added as attributes to the standings, this caused NPE's.  This fixes that.

johnbrvc committed:
i_1116 Fixed ResultsFile bug  found by JUnit

The new method, createCitationRankInformation() in ResultsFile always expected at least as many teams in the standings as there were medals being awarded.  The JUnits pointed out this is necessarily the case.  As such, a safeguard was added to prevent calculation of WF Honors citations if there are less teams than medals in the standings.

johnbrvc committed:
i_1116 Add totalTeams and teamCount attributes to scoreboard XML

To help make better scoreboards using XSL, the totalTeams attribute was added to the contestStandings tag which is the total # of teams that appear on a full scoreboard.
Added teamCount attribute to group tag which is the number of teams in that group.

johnbrvc committed:
i_1116 Ignore new attributes when deserializing XML

JUnit broke due to new attribute added to group tag (teamCount).  Add ignore unknown attributes annotation.

clevengr reviewed (commented):
Reviewed `src/edu/csus/ecs/pc2/api/IProblemDetails.java`.

The class header javadoc for this class is completely wrong -- it was clearly copy/pasted from a different class (and never updated).  I'm aware that the code has been this way for "many a year"... but it would be nice if this PR included a CI that replaced the bogus description with an (even minimal) correct description :)

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithm.java:
```diff
@@ -564,15 +571,38 @@ private void dumpBalloonSettings(BalloonSettings balloonSettings, Problem[] prob
         memento.putInteger("groupCount", groups.length+1);
         IMemento groupsMemento = memento.createChild("groupList");
         int id = 0;
+        int aCount;
```
I'm not crazy about this variable name; it gives no hint as to what is being "counted".  I'd suggest something more verbose, like "teamsInGroupCount".  (I'm not hung up on the particular name; I just think variable names should at least give a hint as to what their purpose is...  also, the name "aCount" seems a bit too close to "account" to me...)

johnbrvc replied:
This was my thinking exactly.  It counts... Accounts, hence, aCount.  I thought it was clever, and, there IS a comment that says what it does:  // Count number of scoreboard displayed teams in this group.   However clever I may find it, evidently I'm the only one :-(   Since the property is called "teamCount" in the XML, I think teamCount is a reasonable name for the counting variable.

Changed in next push.

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithm.java:
```diff
@@ -831,6 +877,46 @@ private void createStandingXML (TreeMap<StandingsRecord, StandingsRecord> treeMa
+        // entire contest, not one particular group (or groups).  So, if any groups were excluded
+        // from the scoreboard, we do not fill in citations.
+        if(!excludedGroups) {
+            // Now go back and fill in the awards for each team if not for specific group
```
I don't understand what "... if not for specific group" means in the context of this comment...

johnbrvc replied:
It was just reiterating what was said in the comment before the "if()".  Specifically, we only fill in the awards if <the scoreboard> is not for a specific group (or groups), that is, it's the entire scoreboard.  Anyway, I removed the last part of the comment and just have it say it is "filling in the award citations ..." since it should be clear from the comment before the "if()" why it's not done all the time.

Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ */
+public class CitationRankInformation {
+    private int lastGoldPlace;
+    private int lastSilverPlace;
+    private int lastBronzePlace;
```
I hate to get bogged down discussions about "Bill terminology", but...  for all the "honors" variables you used "rank" as part of the name (which seems logical).  What doesn't seem logical to me is the use of "place" for the medal variables.  "Bill terminology" aside, it seems like it would be more logical to call these variables `lastGoldRank`, `lastSilverRank`, and `lastBronzeRank`, as that would be consistent with the "lastXXXRank" names of the following variables.

I do think using "rank" makes more sense; I think Bill's rules allow multiple different teams to be in the same "Place" but I think "Rank" is something that's more specific to "monotonically increasing numbers, different for every team".

johnbrvc replied:
Medals are not based on rank, they are based on place, at least that is my understanding.  If there are 4 medals each of G/S/B, then 1st place through 4th place gets gold, 5th place through 8th silver, 9th-12th bronze, independent of the number of problems solved(!).  Rank is based on the number of problem solved.  1-12 are always 1-12.  If the first 4 teams all solve the same # of problems, they all do not have rank 1.  This is different once we are out of the medals.  Again, at least this is my understanding and using "_place_" shows that distinction.   I refer you to this page: [Results for 48th ICPC World Finals](https://icpc.global/worldfinals/results) Note that several teams in the "medals" have the same # of problems solved, yet they do not share the same rank.    Again, this could be Bill terminology.

I'm inclined not to change it.

clevengr replied:
Yeah, this requires a detailed discussion of "Bill's Rules", which I started trying to understand in 1998 and STILL can't grasp.  It's not worth trying to change unless/until we get a clear understanding of it -- and I don't see that happening.  No change necessary...

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,88 @@
+    private int lastHonorsRank;
+    private int firstHonorableMentionRank;
+
+    public int getLastGoldPlace() {
```
...And of course if you accept my suggestion that "rank" should be used for all the award boundary names, this method (and the following ones) should be refactored -- e.g. to `getLastGoldRank`.

johnbrvc replied:
Not inclined to change it (see response above).

clevengr replied:
Yup; same response.

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,88 @@
+        return lastHighestHonorsRank;
+    }
+    /**
+     * Updates highest honors rank if supplied rank is smaller than current one
```
This comment doesn't seem accurate to me -- it appears that what the actual CODE does is "Updates highest honors rank if supplied rank is _greater_ than the current value"

johnbrvc replied:
Oh, yes.  That's wrong.  Comment adjusted.  It's funny I had the other ones right.
Next push.

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,88 @@
+     * @param lastHighestHonorsRank
+     */
+    public void updateLastHighestHonorsRank(int lastHighestHonorsRank) {
+        if(lastHighestHonorsRank > this.lastHighestHonorsRank) {
+            this.lastHighestHonorsRank = lastHighestHonorsRank;
```
The comment above (about the javadoc comment) notwithstanding, I'm not sure I understand the logic behind only allowing a client (i.e., a caller of this method) to only be able to change the rank in one direction.  It seems to me that the boundary between honors divisions could change _in either direction_ depending on (for example) the number of teams in Honorable Mention, or (for example) the number of problems solved by the lowest bronze medalist.  What's the logic behind keeping a client from being able to move the boundary in either direction?  Wouldn't the _client_ be the one who best knows what the 'current boundary' should be?

Note that this same comment applies to the following `updateXXX` methods also...

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,88 @@
+     * Updates high honors rank if supplied rank is bigger than current one
+     * @param lastHighestHonorsRank
+     */
+    public void updateLastHighHonorsRank(int lastHighHonorsRank) {
```
I find the parameter name `lastHighHonorsRank` confusing -- because it is exactly the same as the corresponding field variable which it (potentially) replaces.  (Yes, of course I know it's legal; I just think it's unnecessarily confusing).  I would prefer a name like `newLastHighHonorsRank`, or `proposedLastHighHonorsRank`, or maybe just `newRank` or `proposedRank` for conciseness.

Same comment applies to the other methods (above and below) which use parameter names identical to corresponding field names.

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,88 @@
+        return firstHonorableMentionRank;
+    }
+    /**
+     * Updates highest honorable mention rank if bigger rank is larger than current one
```
I'm not sure I understand this comment -- what does `bigger rank is larger than current one` mean?

clevengr commented on src/edu/csus/ecs/pc2/exports/ccs/ResultsFile.java:
```diff
@@ -281,6 +282,118 @@ public String[] createFileLines(IInternalContest contest, Group group, String re
+     * @param contest
+     * @param standingsRecords array of StandingsRecord (this will be modified and sorted according to the
+     *        current standings sorting rules in effect.  eg. Bill Rules for WF)
+     * @return CititationInformation (ranks for each citation)
```
Was it intentional to not use the actual returned class name in the `return` annotation?

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/exports/ccs/ResultsFile.java:
```diff
@@ -281,6 +282,118 @@ public String[] createFileLines(IInternalContest contest, Group group, String re
+     */
+    public CitationRankInformation createCitationRankInformation(IInternalContest contest, StandingsRecord [] standingsRecords)  {
+
+        int median = getMedian(standingsRecords);
```
Method `getMedian()`, which contains a notation saying that it was "copied from DefaultScoringAlgorithm", appears to be returning the _median number of problems solved_, which it appears is then used (below) to determine the boundary between `HonorableMention` and `Honors`.  However, the WF Rules at https://icpc.global/worldfinals/rules (under "Championship Honors") says:  `Honorable Mention:
All teams solving fewer than the number of problems solved by the median scoring team.`  There is (potentially) a difference between "median number of problems solved" and "number of problems solved by the median team" -- and such a difference would produce a different boundary.

I don't know what the `getMedian()` in `DefaultScoringAlgorithm` is used for (I suspect it is being used erroneously), but it appears to me that `getMedian()` in THIS class needs to find the _median TEAM_ and return the number of problems solved by _that team_.

johnbrvc replied:
I believe the routine does what you say.  The `srArray `is the list of sorted teams.  it goes right to the middle of the list (median) and gets that team's # of problems solved (eg. the number of problems solved by the median team).  If the array has an even number of teams, it averages the # solved by the 2 teams in the middle (floor()).  So, the median it returns is the problem cutoff for non-honorable mention.  Anything less than the value return is HM.

The confusion may be the comment at the top of the routine that says, "What is the median number of problem solved?"  perhaps it should say, "What is the number of problems solved by the median team?" (which is what the code does).
I will fix the comment in both files.

Also verified that the usage of "getMedian()" is correct, and it is.

Comments fixed in next push - no functional changes required.

clevengr commented on src/edu/csus/ecs/pc2/exports/ccs/ResultsFile.java:
```diff
@@ -281,6 +282,118 @@ public String[] createFileLines(IInternalContest contest, Group group, String re
+        Vector<Account> accountVector = contest.getAccounts(Type.TEAM);
+        Account[] accounts = accountVector.toArray(new Account[accountVector.size()]);
+        AccountList accountList = new AccountList();
+        for (Account account : accounts) {
+            accountList.add(account);
+        }
```
Shouldn't this `for-loop` be excluding teams which do not have the `Display-on-scoreboard` property?

johnbrvc replied:
It doesn't matter.  It's just making up a _hashmap_ of accounts, so when it sorts the `standingsRecords` (line 347) the `FinalsStandingsRecordComparator  `can quickly lookup the account to get the team name for sorting.

No changes.

clevengr commented on src/edu/csus/ecs/pc2/core/standings/ProblemSummaryInfo.java:
```diff
@@ -6,14 +6,17 @@
 import javax.xml.bind.annotation.XmlAttribute;
 import javax.xml.bind.annotation.XmlRootElement;
 
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
```
You changed the copyright dates in all the other modified files; I'm surprised you didn't update this one too :)

johnbrvc replied:
Fixed in next push.

clevengr commented on src/edu/csus/ecs/pc2/core/util/RunStatistics.java:
```diff
@@ -16,7 +16,7 @@
 
 /**
  * Contains methods that compute various statistics.
- * 
```
Oh, another file where the copyright date didn't get updated (not that this is in ANY way critical...)

johnbrvc replied:
Fixed in next push.

clevengr reviewed (changes requested):
I reviewed all the files in this PR.  Sorry that I wasn't able to review (and comment on them) in isolation; I found I had to bounce back and forth between the files a lot and GitHub doesn't seem to provide a way to separate "review comments" based on the file to which they're attached.

I had important comments (things which I think need to be addressed -- i.e. at least looked at and responded to -- before I'd be willing to approve the PR) on just the following file:

src/edu/csus/ecs/pc2/exports/ccs/ResultsFile.java


I had minor comments on the following files; I could live with approving the PR without these being addressed:

src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java
src/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithm.java

Proceding now to runtime testing...

clevengr commented on src/edu/csus/ecs/pc2/core/scoring/DefaultScoringAlgorithm.java:
```diff
@@ -831,6 +877,46 @@ private void createStandingXML (TreeMap<StandingsRecord, StandingsRecord> treeMa
+                    // For medals, we use the "place" not the "rank"
+                    // For honors we rank since multiple ranks can be part of each of the honors groups.
+                    if(ri.getFirstHonorableMentionRank() > 0 && teamRank >= ri.getFirstHonorableMentionRank()) {
+                        standingsRecordMemento.putBoolean("isHonable", true);
```
Oops, I found another place where I would request a change before approving:  the text name of the attribute in the above line should be `isHonorable`, not `isHonable`.

johnbrvc replied:
Fixed in next push.

clevengr reviewed (changes requested):
request changing attribute name `isHonable` to `isHonorable` (line 900, class `core/scoring/DefaultScoringAlgorithm.java`

clevengr commented:
I completed a full runtime test of the PR using both the NAC2025 and WF48 (Astana) contest final standings.  I verified that the XML output files for both contests contain the additional (new) attributes as specified in the PR, and that otherwise the results from the PR build match the original (posted) contest results.  Once the comments/issues I raised in at least the two files `exports/ccs/ResultsFile.java` and `core/scoring/DefaultScoringAlgorithm.java` are addressed, I'm ready to approve the PR.

johnbrvc commented:
> Reviewed `src/edu/csus/ecs/pc2/api/IProblemDetails.java`.
> 
> The class header javadoc for this class is completely wrong -- it was clearly copy/pasted from a different class (and never updated). I'm aware that the code has been this way for "many a year"... but it would be nice if this PR included a CI that replaced the bogus description with an (even minimal) correct description :)

WIll be fixed in next push.

johnbrvc committed:
i_1116 PR Change requests

Fixed comments.
Refactored variable name.
Fixed spelling of XML attribute for isHonorable (this is the only functional change in this commit).

clevengr reviewed (approved):
I reviewed the changes made in response to my initial review and request-for-changes.
The changes which I said I felt were required have been made, so I'm ready to approve the PR.

There were a couple of comments I made that I was hoping to get some feedback/explanation about, although they don't rise to the level of withholding approval.  Specifically, in `core/scoring/CitationRankInformation.java`:
- I'm still not sure I understand the reasoning behind only allowing changes to the rank threshholds IN ONE DIRECTION.  It seems to me that the invoking client should have the ability to either raise OR LOWER the threshholds.
- there is a comment (line 80) that says: `Updates highest honorable mention rank if bigger rank is larger than current one`.  I'm not at all sure what that really means...

That said, I approve the PR.

johnbrvc committed:
i_1116 PR change request on comments

Updated comments in CitationRankInformation to reflect what is actually happening.

johnbrvc commented:
> * I'm still not sure I understand the reasoning behind only allowing changes to the rank threshholds IN ONE DIRECTION.  It seems to me that the invoking client should have the ability to either raise OR LOWER the threshholds.

The CitationRankInformation class keeps track of the cut-offs for each of the "honors" ranks.  Once a rank, for say, high honors is set by team, any other team with a lower rank for that honors category should not modify it; the idea is to keep track of the highest rank for each honors category; if we allow a team with a lower rank to overwrite a higher one, the class is useless.    That's why the methods are called "updateLastXXXXXXRank()" instead of "setLastXXXXXXRank()".

> * there is a comment (line 80) that says: `Updates highest honorable mention rank if bigger rank is larger than current one`.  I'm not at all sure what that really means...
> 
That comment is wrong.  Honorable Mention works the other way.  We want the SMALLEST rank for honorable mention, not the highest.  Basically, all teams with that rank and greater are HM.  Whereas for the Bill honors ranks, we want the highest (biggest) rank for each honors category.  Comment updated.
 
> That said, I approve the PR.

JoeTerlizzi commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,91 @@
+    }
+    /**
+     * Updates high honors rank if supplied rank is bigger than current one
+     * @param lastHighestHonorsRank
```
Should be "@param lastHighHonorsRank"?

johnbrvc replied:
Fixed in next push.   Please resolve this conversation.

JoeTerlizzi commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,91 @@
+    }
+    /**
+     * Updates honors rank if supplied rank is bigger than current one
+     * @param lastHighestHonorsRank
```
Should be "@param lastHonorsRank"?

johnbrvc replied:
Fixed in next push.  Please resolve this conversation.

JoeTerlizzi commented on src/edu/csus/ecs/pc2/core/scoring/CitationRankInformation.java:
```diff
@@ -0,0 +1,91 @@
+     * This keeps track of the lowest rank for HM so all teams with this rank and
+     * higher are HM.
+     *
+     * @param lastHighestHonorsRank
```
Should be "@param firstHonorableMentionRank"?

johnbrvc replied:
Fixed in next push.    Please resolve this conversation.

JoeTerlizzi reviewed (approved):
Tested with NAC2025.  Left some doc comments to fix.  Approved.

johnbrvc committed:
i_1116 PR Change requests

Fixed some wrong @param javadoc comments.
